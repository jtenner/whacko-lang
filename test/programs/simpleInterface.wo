interface Test {
  test(): i32;
}

// I'll do it
class A implements Test {
  constructor() {}

  test(): i32 {
    return i32(42);
  }
}

class B implements Test {
  constructor() {}

  test(): i32 {
    return i32(69);
  }
}

class C implements Test {
  constructor() {}

  test(): i32 {
    return i32(123);
  }
}

class D implements Test {
  constructor() {}

  test(): i32 {
    return i32(456);
  }
}

export fn callTest(myTest: Test): i32 {
  return myTest.test();
}

export fn _start(): void {
  let a = new A();
  let b = new B();
  let c = new C();
  let d = new D();

  assert(callTest(a) == i32(42));
  assert(callTest(b) == i32(69));
  assert(callTest(c) == i32(123));
  assert(callTest(d) == i32(456));
}
