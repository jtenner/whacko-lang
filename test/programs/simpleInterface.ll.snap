; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

declare ptr @__whacko_gc_alloc(i32, i32)

declare void @__whacko_gc_store_field_ptr(i32, ptr, ptr, ptr)

define void @_start() {
entry:
  %"alloca~0" = alloca ptr, align 8
  %"alloca~1" = alloca ptr, align 8
  %"alloca~2" = alloca ptr, align 8
  %"alloca~3" = alloca ptr, align 8
  %"result~227" = call ptr @__whacko_gc_alloc(i32 1, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/simpleInterface.wo]A.constructor;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>voidType"(ptr %"result~227")
  store ptr %"result~227", ptr %"alloca~0", align 8
  %"result~229" = call ptr @__whacko_gc_alloc(i32 2, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/simpleInterface.wo]B.constructor;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>voidType"(ptr %"result~229")
  store ptr %"result~229", ptr %"alloca~1", align 8
  %"result~231" = call ptr @__whacko_gc_alloc(i32 3, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/simpleInterface.wo]C.constructor;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>voidType"(ptr %"result~231")
  store ptr %"result~231", ptr %"alloca~2", align 8
  %"result~233" = call ptr @__whacko_gc_alloc(i32 4, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/simpleInterface.wo]D.constructor;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>voidType"(ptr %"result~233")
  store ptr %"result~233", ptr %"alloca~3", align 8
  %"101~inst" = load ptr, ptr %"alloca~0", align 8
  %"105~inst" = call i32 @callTest(ptr %"101~inst")
  %"111~inst" = icmp eq i32 %"105~inst", 42
  br i1 %"111~inst", label %"next~114", label %"falsy~113"

"falsy~113":                                      ; preds = %entry
  unreachable

"next~114":                                       ; preds = %entry
  %"118~inst" = load ptr, ptr %"alloca~1", align 8
  %"121~inst" = call i32 @callTest(ptr %"118~inst")
  %"127~inst" = icmp eq i32 %"121~inst", 69
  br i1 %"127~inst", label %"next~130", label %"falsy~129"

"falsy~129":                                      ; preds = %"next~114"
  unreachable

"next~130":                                       ; preds = %"next~114"
  %"134~inst" = load ptr, ptr %"alloca~2", align 8
  %"137~inst" = call i32 @callTest(ptr %"134~inst")
  %"143~inst" = icmp eq i32 %"137~inst", 123
  br i1 %"143~inst", label %"next~146", label %"falsy~145"

"falsy~145":                                      ; preds = %"next~130"
  unreachable

"next~146":                                       ; preds = %"next~130"
  %"150~inst" = load ptr, ptr %"alloca~3", align 8
  %"153~inst" = call i32 @callTest(ptr %"150~inst")
  %"159~inst" = icmp eq i32 %"153~inst", 456
  br i1 %"159~inst", label %"next~162", label %"falsy~161"

"falsy~161":                                      ; preds = %"next~146"
  unreachable

"next~162":                                       ; preds = %"next~146"
  ret void
}

define void @"callable:[test/programs/simpleInterface.wo]A.constructor;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret void
}

define void @"callable:[test/programs/simpleInterface.wo]B.constructor;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret void
}

define void @"callable:[test/programs/simpleInterface.wo]C.constructor;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret void
}

define void @"callable:[test/programs/simpleInterface.wo]D.constructor;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret void
}

define i32 @callTest(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"168~inst" = load ptr, ptr %"alloca~0", align 8
  %"172~inst" = call i32 @"callable:[test/programs/simpleInterface.wo]Test.test;methodType:interfaceType:[test/programs/simpleInterface.wo]Test<>#()=>i32Type"(ptr %"168~inst")
  ret i32 %"172~inst"
}

define i32 @"callable:[test/programs/simpleInterface.wo]Test.test;methodType:interfaceType:[test/programs/simpleInterface.wo]Test<>#()=>i32Type"(ptr %0) {
"entry~234":
  %"typeGEP~235" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~235", align 4
  switch i32 %1, label %"default~236" [
    i32 1, label %"callable:[test/programs/simpleInterface.wo]A.test;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>i32Type"
    i32 2, label %"callable:[test/programs/simpleInterface.wo]B.test;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>i32Type"
    i32 3, label %"callable:[test/programs/simpleInterface.wo]C.test;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>i32Type"
    i32 4, label %"callable:[test/programs/simpleInterface.wo]D.test;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>i32Type"
  ]

"default~236":                                    ; preds = %"entry~234"
  unreachable

"callable:[test/programs/simpleInterface.wo]A.test;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>i32Type": ; preds = %"entry~234"
  %2 = call i32 @"callable:[test/programs/simpleInterface.wo]A.test;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>i32Type"(ptr %0)
  ret i32 %2

"callable:[test/programs/simpleInterface.wo]B.test;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>i32Type": ; preds = %"entry~234"
  %3 = call i32 @"callable:[test/programs/simpleInterface.wo]B.test;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>i32Type"(ptr %0)
  ret i32 %3

"callable:[test/programs/simpleInterface.wo]C.test;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>i32Type": ; preds = %"entry~234"
  %4 = call i32 @"callable:[test/programs/simpleInterface.wo]C.test;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>i32Type"(ptr %0)
  ret i32 %4

"callable:[test/programs/simpleInterface.wo]D.test;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>i32Type": ; preds = %"entry~234"
  %5 = call i32 @"callable:[test/programs/simpleInterface.wo]D.test;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>i32Type"(ptr %0)
  ret i32 %5
}

define i32 @"callable:[test/programs/simpleInterface.wo]A.test;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>i32Type"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret i32 42
}

define i32 @"callable:[test/programs/simpleInterface.wo]B.test;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>i32Type"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret i32 69
}

define i32 @"callable:[test/programs/simpleInterface.wo]C.test;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>i32Type"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret i32 123
}

define i32 @"callable:[test/programs/simpleInterface.wo]D.test;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>i32Type"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  ret i32 456
}

declare void @__whacko_gc_visit(ptr)

define void @__visit_whacko_object(ptr %0) {
"entry~237":
  %"typeGEP~239" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~239", align 4
  switch i32 %1, label %"unreachable~238" [
    i32 1, label %"classType:[test/programs/simpleInterface.wo]A<>"
    i32 2, label %"classType:[test/programs/simpleInterface.wo]B<>"
    i32 3, label %"classType:[test/programs/simpleInterface.wo]C<>"
    i32 4, label %"classType:[test/programs/simpleInterface.wo]D<>"
    i32 5, label %"classType:[std/String.wo]String<>"
  ]

"unreachable~238":                                ; preds = %"entry~237"
  unreachable

"classType:[test/programs/simpleInterface.wo]A<>": ; preds = %"entry~237"
  ret void

"classType:[test/programs/simpleInterface.wo]B<>": ; preds = %"entry~237"
  ret void

"classType:[test/programs/simpleInterface.wo]C<>": ; preds = %"entry~237"
  ret void

"classType:[test/programs/simpleInterface.wo]D<>": ; preds = %"entry~237"
  ret void

"classType:[std/String.wo]String<>":              ; preds = %"entry~237"
  ret void
}
